### PR TITLE
Moved cluster and service to tf12

### PR DESCRIPTION
**Pull request by lukaspour**
Hi @itsdalmo
I am currently transforming Neo to TF12 and this is a very important module for us. I haven't touched `spotfleet-cluster` submodule. Could you take a look at the rest of the code? Also, there is a part with a dynamic block which looks ugly:
```
  health_check = [{
    port    = "traffic-port"
    path    = "/"
    matcher = "200"

    healthy_threshold   = null
    interval            = null
    protocol            = null
    timeout             = null
    unhealthy_threshold = null
  }]
```
Did you also have to deal with these things? Any suggestion would be very welcomed.

### CODE REVIEW

File: examples/default/example.tf [hunk 1]
```diff
@@ -108,13 +108,13 @@ module "application" {
   source = "../../modules/service"
 
   name_prefix                       = "example-app"
-  vpc_id                            = "${data.aws_vpc.main.id}"
-  cluster_id                        = "${module.cluster.id}"
-  cluster_role_name                 = "${module.cluster.role_name}"
-  desired_count                     = "1"
+  vpc_id                            = data.aws_vpc.main.id
+  cluster_id                        = module.cluster.id
+  cluster_role_name                 = module.cluster.role_namr
```
Comment (mikael-lindstrom) [inline examples/default/example.tf]: Small typo here

File: modules/cluster/variables.tf [hunk 2]
```diff
@@ -17,7 +17,7 @@ variable "vpc_id" {
 
 variable "subnet_ids" {
   description = "ID of subnets where instances can be provisioned."
-  type        = "list"
+  type        = list(string)
```
Comment (mikael-lindstrom) [inline modules/cluster/variables.tf]: I think we should add type to all variables for clarity.

**Comment by mikael-lindstrom** [comment]
Some changes are needed to the `Makefile` due to 0.12.

```
terraform validate -check-variables=false; \
```
Needs to be changed to (in both places):
```
AWS_DEFAULT_REGION=eu-west-1 terraform validate; \
```

**Comment by lukaspour** [comment]
I did, it didn't help much

File: Makefile [hunk 3]
```diff
@@ -15,7 +15,7 @@ test:
 	@for d in $$(find . -type f -name '*.tf' -path "./modules/*" -not -path "**/.terraform/*" -exec dirname {} \; | sort -u); do \
 		cd $$d; \
 		terraform init -backend=false >> /dev/null; \
-		terraform validate -check-variables=false; \
+		AWS_DEFAULT_REGION=eu-west-1 terraform validate -check-variables=false; \
```
Comment (mikael-lindstrom) [inline Makefile]: `-check-variables=false` needs to be removed (option removed in version 0.12).

File: modules/cluster/main.tf [hunk 4]
```diff
@@ -74,31 +74,31 @@ data "aws_iam_policy_document" "permissions" {
...
-  ebs_block_devices    = "${var.ebs_block_devices}"
-  tags                 = "${var.tags}"
+  instance_policy      = data.aws_iam_policy_document.permissions.json
+  min_size             = var.min_size
+  max_size             = var.max_size
+  instance_type        = var.instance_type
+  instance_ami         = var.instance_ami
+  instance_key         = var.instance_key
+  instance_volume_size = var.instance_volume_size
+  ebs_block_devices    = var.ebs_block_devices
```
Comment (mikael-lindstrom) [inline modules/cluster/main.tf]: New ASG module expects this to be `list(map(string))` [here](https://github.com/telia-oss/terraform-aws-asg/blob/master/variables.tf#L56)

**Comment by lukaspour** [comment]
I had to rework spotfleet as well to pass the CI tests

**Pull request merged**